src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,533 @@
+	// This is expected to be the same as the CertificateVerify.algorithm
+	// sent by the server.
+	expCertVerfAlgo SignatureScheme
+	// The credential's public key.
+	PublicKey crypto.PublicKey
+}
+
+// DelegatedCredential stores the credential and its signature.
+type DelegatedCredential struct {
+	// The serialized form of the Credential
+	Raw []byte
+
+	// Credential stores the public components of a Delegated Credential
+	Cred *Credential
+
+	// The signature scheme used to sign the credential
+	Algorithm SignatureScheme
+
+	// The Credential's delegation: a signature that binds the credential to
+	// the end-entity certificate's public key
+	Signature []byte
+}
+
+// marshalPublicKeyInfo returns a DER encoded PublicKeyInfo
+// from a Delegated Credential (as defined in the X.509 standard).
+// The following key types are currently supported: *ecdsa.PublicKey
+// and ed25519.PublicKey. Unsupported key types result in an error.
+// rsa.PublicKey is not supported a defined by the draft.
+func (cred *Credential) marshalPublicKeyInfo() ([]byte, error) {
+	switch cred.expCertVerfAlgo {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512,
+		Ed25519:
+		serPub, err := x509.MarshalPKIXPublicKey(cred.PublicKey)
+		if err != nil {
+			return nil, err
+		}
+
+		return serPub, nil
+	default:
+		return nil, fmt.Errorf("tls: unsupported signature scheme: 0x%04x", cred.expCertVerfAlgo)
+	}
+}
+
+// unmarshalPublicKeyInfo parses a DER encoded PublicKeyInfo from a Delegated
+// Credential to a public key and its corresponding algorithm.
+func unmarshalPublicKeyInfo(serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
+	pubKey, err := x509.ParsePKIXPublicKey(serialized)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch pk := pubKey.(type) {
+	case *ecdsa.PublicKey:
+		curveName := pk.Curve.Params().Name
+		if curveName == "P-256" {
+			return pk, ECDSAWithP256AndSHA256, nil
+		} else if curveName == "P-384" {
+			return pk, ECDSAWithP384AndSHA384, nil
+		} else if curveName == "P-521" {
+			return pk, ECDSAWithP521AndSHA512, nil
+		} else {
+			return nil, 0, fmt.Errorf("tls: unsupported delgation key type: %s", curveName)
+		}
+	case ed25519.PublicKey:
+		return pk, Ed25519, nil
+	default:
+		return nil, 0, fmt.Errorf("tls: unsupported delgation key type: %T", pk)
+	}
+}
+
+// marshal encodes the credential struct of the Delegated Credential.
+func (cred *Credential) marshal() ([]byte, error) {
+	ser := make([]byte, 4)
+	binary.BigEndian.PutUint32(ser, uint32(cred.validTime/time.Second))
+
+	var serAlgo [2]byte
+	binary.BigEndian.PutUint16(serAlgo[:], uint16(cred.expCertVerfAlgo))
+	ser = append(ser, serAlgo[:]...)
+
+	// Encode the public key
+	serPub, err := cred.marshalPublicKeyInfo()
+	if err != nil {
+		return nil, err
+	}
+	// Assert that the public key encoding is no longer than 2^24  bytes.
+	if len(serPub) > dcMaxPubLen {
+		return nil, errors.New("tls: public key length exceeds 2^24-1 limit")
```

Comment:
dcMaxPubLen is 2^24, so 2^24 is allowed which is against the RFC which only allows up to 2^24-1.